src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -1,401 +1,427 @@
+    if (dir->update_field(&victim, "globalWeight", std::to_string(victim.globalWeight), y) < 0) {
+      return -1;
+    }
+
+    if (dir->remove_host(&victim, dir->cct->_conf->rgw_local_cache_address, y) < 0) {
+      return -1;
+    } else {
+      if (cacheDriver->del(dpp, key, y) < 0) {
+        return -1;
+      } else {
+	ldpp_dout(dpp, 10) << "RGW D4N Policy: Block " << victim.cacheObj.objName << " has been evicted." << dendl;
+
+	uint64_t num_entries = entries_map.size();
+
+	if (!avgWeight) {
+	  if (set_min_avg_weight(0, dir->cct->_conf->rgw_local_cache_address, y) < 0) // Where else must this be set? -Sam 
+	    return -1;
+	} else {
+	  if (set_min_avg_weight(avgWeight - (it->second->localWeight/num_entries), dir->cct->_conf->rgw_local_cache_address, y) < 0) // Where else must this be set? -Sam 
+	    return -1;
+	} 
+
+	int age = get_age(y);
+	age = std::max(it->second->localWeight, age);
+	if (set_age(age, y) < 0)
+	  return -1;
+      }
     }
-  } 
 
+    freeSpace = cacheDriver->get_free_space(dpp);
+  }
+  
   return 0;
 }
 
-uint64_t LFUDAPolicy::eviction(const DoutPrefixProvider* dpp, rgw::cache::CacheDriver* cacheNode) {
-  CacheBlock victim = find_victim(dpp, cacheNode);
-
-  if (victim.cacheObj.objName.empty()) {
-    ldpp_dout(dpp, 10) << "RGW D4N Policy: Could not find victim block" << dendl;
-    return -1;
+void LFUDAPolicy::update(const DoutPrefixProvider* dpp, std::string& key, uint64_t offset, uint64_t len, std::string version, optional_yield y)
+{
+  using handle_type = boost::heap::fibonacci_heap<LFUDAEntry*, boost::heap::compare<EntryComparator<LFUDAEntry>>>::handle_type;
+
+  int age = get_age(y); 
+  int localWeight = age;
+  auto entry = find_entry(key);
+  if (entry != nullptr) { 
+    entry->localWeight += age;
+    localWeight = entry->localWeight;
+  }  
+
+  erase(dpp, key);
+  
+  LFUDAEntry *e = new LFUDAEntry(key, offset, len, version, localWeight);
+  handle_type handle = entries_heap.push(e);
+  e->set_handle(handle);
+  entries_map.emplace(key, e);
+
+  if (cacheDriver->set_attr(dpp, key, "localWeight", std::to_string(localWeight), y) < 0) {
+    ldpp_dout(dpp, 10) << "LFUDAPolicy::update:: " << __func__ << "(): Cache driver set_attr method failed." << dendl;
   }
+}
 
-  std::string key = "rgw-object:" + victim.cacheObj.objName + ":directory";
-  std::string hosts;
-  int globalWeight = get_global_weight(key);
-  int localWeight = std::stoi(cacheNode->get_attr(dpp, victim.cacheObj.objName, "localWeight")); // change to block name eventually -Sam
-  int avgWeight = get_min_avg_weight();
-
-  if (exist_key(key)) {
-    try {
-      client.hget(key, "hostsList", [&hosts](cpp_redis::reply& reply) {
-	if (!reply.is_null()) {
-	  hosts = reply.as_string();
-	}
-      });
-
-      client.sync_commit(std::chrono::milliseconds(1000));
-    } catch(std::exception &e) {
-      return -1;
+bool LFUDAPolicy::erase(const DoutPrefixProvider* dpp, const std::string& key)
+{
+  for (auto const& it : entries_heap) {
```

Comment:
To avoid iterating through the entire min heap, can you do the reverse - look up the entry in the map, and then use the entry address to erase it from the heap also - like it is done in LRUPolicy::erase()

Reply:
Also, you will have to use locks in `update` and `erase` (similar to `LRUPolicy::erase` and `update`).

Reply:
Is there anywhere else I need the lock? I see `LRUPolicy` also uses it in `eviction`

Reply:
Check `key_exists` also. Basically any method that accesses the data structures.